File: amie-decoder/src/main/java/org/apache/custos/amie/service/PersonService.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.custos.amie.service;

import com.fasterxml.jackson.databind.JsonNode;
import org.apache.custos.amie.model.ClusterAccountEntity;
import org.apache.custos.amie.model.PersonDnsEntity;
import org.apache.custos.amie.model.PersonEntity;
import org.apache.custos.amie.repo.ClusterAccountRepository;
import org.apache.custos.amie.repo.PersonDnsRepository;
import org.apache.custos.amie.repo.PersonRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.Assert;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Service responsible for managing the lifecycle of Person entities and their associated DNs.
 */
@Service
public class PersonService {

    private static final Logger LOGGER = LoggerFactory.getLogger(PersonService.class);

    private final PersonRepository personRepository;
    private final PersonDnsRepository personDnsRepository;
    private final ClusterAccountRepository clusterAccountRepository;

    public PersonService(PersonRepository personRepository, PersonDnsRepository personDnsRepository, ClusterAccountRepository clusterAccountRepository) {
        this.personRepository = personRepository;
        this.personDnsRepository = personDnsRepository;
        this.clusterAccountRepository = clusterAccountRepository;
    }

    /**
     * Finds a Person by their AMIE Global ID or creates a new one if not found.
     *
     * @param packetBody The "body" of the AMIE packet containing user details.
     * @return The existing or newly created PersonEntity.
     */
    @Transactional
    public PersonEntity findOrCreatePersonFromPacket(JsonNode packetBody) {
        String accessGlobalId = packetBody.path("UserGlobalID").asText();
        Assert.hasText(accessGlobalId, "Packet body must contain a 'UserGlobalID'.");

        Optional<PersonEntity> existingPerson = personRepository.findByAccessGlobalId(accessGlobalId);
        if (existingPerson.isPresent()) {
            LOGGER.info("Found existing person with local ID [{}] for access_global_id [{}]", existingPerson.get().getId(), accessGlobalId);
            return existingPerson.get();
        }

        // If not found, create a new person
        LOGGER.info("No person found for access_global_id [{}]. Creating a new person record.", accessGlobalId);
        PersonEntity newPerson = new PersonEntity();
        newPerson.setId(UUID.randomUUID().toString());
        newPerson.setAccessGlobalId(accessGlobalId);
        newPerson.setFirstName(packetBody.path("UserFirstName").asText(""));
        newPerson.setLastName(packetBody.path("UserLastName").asText(""));
        newPerson.setEmail(packetBody.path("UserEmail").asText(""));
        newPerson.setOrganization(packetBody.path("UserOrganization").asText(null));
        newPerson.setOrgCode(packetBody.path("UserOrgCode").asText(null));
        newPerson.setNsfStatusCode(packetBody.path("NsfStatusCode").asText(null));
        personRepository.save(newPerson);

        // Save their associated DNs
        JsonNode dnList = packetBody.path("UserDnList");
        if (dnList.isArray()) {
            for (JsonNode dnNode : dnList) {
                String dn = dnNode.asText(null);
                if (dn != null && !dn.isBlank()) {
                    PersonDnsEntity pde = new PersonDnsEntity();
                    pde.setPerson(newPerson);
                    pde.setDn(dn);
                    personDnsRepository.save(pde);
                }
            }
        }
        return newPerson;
    }


    @Transactional
    public void replaceFromModifyPacket(JsonNode body) {
        String personId = body.path("PersonID").asText(null);
        Assert.hasText(personId, "Missing required 'PersonID' in request_user_modify replace body");

        PersonEntity person = personRepository.findById(personId)
                .orElseThrow(() -> new IllegalArgumentException("Unknown local PersonID: " + personId));

        // Update fields
        if (body.has("FirstName")) person.setFirstName(body.path("FirstName").asText(person.getFirstName()));
        if (body.has("LastName")) person.setLastName(body.path("LastName").asText(person.getLastName()));
        if (body.has("Email")) person.setEmail(body.path("Email").asText(person.getEmail()));
        person.setOrganization(body.has("Organization") ? body.path("Organization").asText(null) : null);
        person.setOrgCode(body.has("OrgCode") ? body.path("OrgCode").asText(null) : null);
        person.setNsfStatusCode(body.has("NsfStatusCode") ? body.path("NsfStatusCode").asText(null) : null);
        personRepository.save(person);

        Set<String> newDns = new HashSet<>();
        JsonNode dnList = body.path("DnList");
        if (dnList != null && dnList.isArray()) {
            for (JsonNode dnNode : dnList) {
                String dn = dnNode.asText(null);
                if (dn != null && !dn.isBlank()) newDns.add(dn);
            }
        }

        if (newDns.isEmpty()) {
            personDnsRepository.deleteByPerson_Id(personId);
        } else {
            personDnsRepository.deleteByPerson_IdAndDnNotIn(personId, new ArrayList<>(newDns));
            for (String dn : newDns) {
                if (!personDnsRepository.existsByPerson_IdAndDn(personId, dn)) {
                    PersonDnsEntity p = new PersonDnsEntity();
                    p.setPerson(person);
                    p.setDn(dn);
                    personDnsRepository.save(p);
                }
            }
        }
    }

    @Transactional
    public void deleteFromModifyPacket(JsonNode body) {
        String personId = body.path("PersonID").asText(null);
        Assert.hasText(personId, "Missing required 'PersonID' in request_user_modify delete body");

        // Cascades will remove DNs and cluster accounts
        personRepository.deleteById(personId);
    }

    @Transactional
    public void mergePersons(String survivingPersonId, String retiringPersonId) {
        LOGGER.info("Merging person {} into {}", retiringPersonId, survivingPersonId);

        PersonEntity survivingPerson = personRepository.findById(survivingPersonId)
                .orElseThrow(() -> new IllegalStateException("Could not find surviving person with local ID: " + survivingPersonId));

        PersonEntity retiringPerson = personRepository.findById(retiringPersonId)
                .orElseThrow(() -> new IllegalStateException("Could not find retiring person with local ID: " + retiringPersonId));

        // Re-associate all cluster accounts
        for (ClusterAccountEntity account : retiringPerson.getClusterAccounts()) {
            LOGGER.info("Moving cluster account '{}' from retiring person to surviving person", account.getUsername());
            account.setPerson(survivingPerson);
            clusterAccountRepository.save(account);
        }

        // Merge DNs, avoiding duplicates
        Set<String> survivingDns = survivingPerson.getDnsEntries().stream()
                .map(PersonDnsEntity::getDn)
                .collect(Collectors.toSet());

        for (PersonDnsEntity retiringDn : retiringPerson.getDnsEntries()) {
            if (!survivingDns.contains(retiringDn.getDn())) {
                LOGGER.info("Moving DN '{}' from retiring person to surviving person", retiringDn.getDn());
                retiringDn.setPerson(survivingPerson);
                personDnsRepository.save(retiringDn);
            } else {
                personDnsRepository.delete(retiringDn);
            }
        }

        // The CASCADE constraint will clean up any remaining associations
        personRepository.delete(retiringPerson);
        LOGGER.info("Successfully merged and deleted retiring person record {}", retiringPersonId);
    }
}

File: allocations/access-amie/service/person_service.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package service

import (
	"context"
	"database/sql"
	"fmt"
	"log/slog"

	"github.com/apache/airavata-custos/allocations/access-amie/model"
	"github.com/google/uuid"
)

type personStore interface {
	FindByID(ctx context.Context, id string) (*model.Person, error)
	FindByAccessGlobalID(ctx context.Context, globalID string) (*model.Person, error)
	Save(ctx context.Context, tx *sql.Tx, p *model.Person) error
	Delete(ctx context.Context, tx *sql.Tx, id string) error
}

type personDNStore interface {
	ExistsByPersonAndDN(ctx context.Context, personID, dn string) (bool, error)
	Save(ctx context.Context, tx *sql.Tx, d *model.PersonDN) error
	DeleteByPersonID(ctx context.Context, tx *sql.Tx, personID string) error
	DeleteByPersonIDNotIn(ctx context.Context, tx *sql.Tx, personID string, dnsToKeep []string) error
	FindByPersonID(ctx context.Context, personID string) ([]model.PersonDN, error)
}

type personAccountStore interface {
	FindByPerson(ctx context.Context, personID string) ([]model.ClusterAccount, error)
	UpdatePersonID(ctx context.Context, tx *sql.Tx, accountID, newPersonID string) error
}

type PersonService struct {
	persons  personStore
	dns      personDNStore
	accounts personAccountStore
}

func NewPersonService(persons personStore, dns personDNStore, accounts personAccountStore) *PersonService {
	return &PersonService{
		persons:  persons,
		dns:      dns,
		accounts: accounts,
	}
}

// FindOrCreateFromPacket looks up a person by their ACCESS Global ID or
// creates a new person record from the supplied AMIE packet body.
func (s *PersonService) FindOrCreateFromPacket(ctx context.Context, tx *sql.Tx, body map[string]any) (*model.Person, error) {
	globalID, _ := body["UserGlobalID"].(string)
	if globalID == "" {
		return nil, fmt.Errorf("person_service: UserGlobalID is required")
	}

	existing, err := s.persons.FindByAccessGlobalID(ctx, globalID)
	if err != nil {
		return nil, fmt.Errorf("person_service: finding person by global ID %s: %w", globalID, err)
	}
	if existing != nil {
		return existing, nil
	}

	firstName, _ := body["UserFirstName"].(string)
	lastName, _ := body["UserLastName"].(string)
	email, _ := body["UserEmail"].(string)

	p := &model.Person{
		ID:             uuid.NewString(),
		AccessGlobalID: globalID,
		FirstName:      firstName,
		LastName:       lastName,
		Email:          email,
		Organization:   optionalString(body, "UserOrganization"),
		OrgCode:        optionalString(body, "UserOrgCode"),
		NsfStatusCode:  optionalString(body, "NsfStatusCode"),
	}

	if err := s.persons.Save(ctx, tx, p); err != nil {
		return nil, fmt.Errorf("person_service: saving new person %s: %w", p.ID, err)
	}

	slog.DebugContext(ctx, "created person from packet", "person_id", p.ID, "global_id", globalID)

	// Persist DN list if present.
	if dnList, ok := body["UserDnList"].([]any); ok {
		for _, raw := range dnList {
			dn, _ := raw.(string)
			if dn == "" {
				continue
			}
			d := &model.PersonDN{PersonID: p.ID, DN: dn}
			if err := s.dns.Save(ctx, tx, d); err != nil {
				return nil, fmt.Errorf("person_service: saving DN for person %s: %w", p.ID, err)
			}
		}
	}

	return p, nil
}

// ReplaceFromModifyPacket selectively updates person fields that are present
// in the AMIE modify-person packet body.
func (s *PersonService) ReplaceFromModifyPacket(ctx context.Context, tx *sql.Tx, body map[string]any) error {
	personID, _ := body["PersonID"].(string)
	if personID == "" {
		return fmt.Errorf("person_service: PersonID is required for modify")
	}

	p, err := s.persons.FindByID(ctx, personID)
	if err != nil {
		return fmt.Errorf("person_service: finding person %s: %w", personID, err)
	}
	if p == nil {
		return fmt.Errorf("person_service: person %s not found", personID)
	}

	// Update only fields present in the body.
	if v, ok := body["UserFirstName"]; ok {
		p.FirstName, _ = v.(string)
	}
	if v, ok := body["UserLastName"]; ok {
		p.LastName, _ = v.(string)
	}
	if v, ok := body["UserEmail"]; ok {
		p.Email, _ = v.(string)
	}
	if _, ok := body["UserOrganization"]; ok {
		p.Organization = optionalString(body, "UserOrganization")
	}
	if _, ok := body["UserOrgCode"]; ok {
		p.OrgCode = optionalString(body, "UserOrgCode")
	}
	if _, ok := body["NsfStatusCode"]; ok {
		p.NsfStatusCode = optionalString(body, "NsfStatusCode")
	}

	// Handle DN list updates.
	if rawDNs, ok := body["UserDnList"]; ok {
		dnList, isList := rawDNs.([]any)
		if isList && len(dnList) > 0 {
			var dnsToKeep []string
			for _, raw := range dnList {
				dn, _ := raw.(string)
				if dn != "" {
					dnsToKeep = append(dnsToKeep, dn)
				}
			}
			if err := s.dns.DeleteByPersonIDNotIn(ctx, tx, personID, dnsToKeep); err != nil {
				return fmt.Errorf("person_service: pruning DNs for person %s: %w", personID, err)
			}
			for _, dn := range dnsToKeep {
				exists, err := s.dns.ExistsByPersonAndDN(ctx, personID, dn)
				if err != nil {
					return fmt.Errorf("person_service: checking DN existence for person %s: %w", personID, err)
				}
				if !exists {
					d := &model.PersonDN{PersonID: personID, DN: dn}
					if err := s.dns.Save(ctx, tx, d); err != nil {
						return fmt.Errorf("person_service: saving DN for person %s: %w", personID, err)
					}
				}
			}
		} else if isList && len(dnList) == 0 {
			if err := s.dns.DeleteByPersonID(ctx, tx, personID); err != nil {
				return fmt.Errorf("person_service: clearing DNs for person %s: %w", personID, err)
			}
		}
	}

	if err := s.persons.Save(ctx, tx, p); err != nil {
		return fmt.Errorf("person_service: saving updated person %s: %w", personID, err)
	}

	slog.DebugContext(ctx, "updated person from modify packet", "person_id", personID)
	return nil
}

// DeleteFromModifyPacket removes a person identified by the packet body.
// Cascade rules in the database handle related cleanup.
func (s *PersonService) DeleteFromModifyPacket(ctx context.Context, tx *sql.Tx, body map[string]any) error {
	personID, _ := body["PersonID"].(string)
	if personID == "" {
		return fmt.Errorf("person_service: PersonID is required for delete")
	}

	if err := s.persons.Delete(ctx, tx, personID); err != nil {
		return fmt.Errorf("person_service: deleting person %s: %w", personID, err)
	}

	slog.DebugContext(ctx, "deleted person from modify packet", "person_id", personID)
	return nil
}

// MergePersons transfers all accounts and DNs from the retiring person to the
// surviving person, then deletes the retiring person.
func (s *PersonService) MergePersons(ctx context.Context, tx *sql.Tx, survivingID, retiringID string) error {
	surviving, err := s.persons.FindByID(ctx, survivingID)
	if err != nil {
		return fmt.Errorf("person_service: finding surviving person %s: %w", survivingID, err)
	}
	if surviving == nil {
		return fmt.Errorf("person_service: surviving person %s not found", survivingID)
	}

	retiring, err := s.persons.FindByID(ctx, retiringID)
	if err != nil {
		return fmt.Errorf("person_service: finding retiring person %s: %w", retiringID, err)
	}
	if retiring == nil {
		return fmt.Errorf("person_service: retiring person %s not found", retiringID)
	}

	// Move cluster accounts from the retiring person to the surviving person.
	retiringAccounts, err := s.accounts.FindByPerson(ctx, retiringID)
	if err != nil {
		return fmt.Errorf("person_service: finding accounts for retiring person %s: %w", retiringID, err)
	}
	for _, acct := range retiringAccounts {
		if err := s.accounts.UpdatePersonID(ctx, tx, acct.ID, survivingID); err != nil {
			return fmt.Errorf("person_service: moving account %s to surviving person %s: %w", acct.ID, survivingID, err)
		}
	}

	// Merge DNs: copy any DNs from the retiring person that the surviving
	// person does not already have.
	retiringDNs, err := s.dns.FindByPersonID(ctx, retiringID)
	if err != nil {
		return fmt.Errorf("person_service: finding DNs for retiring person %s: %w", retiringID, err)
	}
	for _, dn := range retiringDNs {
		exists, err := s.dns.ExistsByPersonAndDN(ctx, survivingID, dn.DN)
		if err != nil {
			return fmt.Errorf("person_service: checking DN existence for surviving person %s: %w", survivingID, err)
		}
		if !exists {
			d := &model.PersonDN{PersonID: survivingID, DN: dn.DN}
			if err := s.dns.Save(ctx, tx, d); err != nil {
				return fmt.Errorf("person_service: saving merged DN for surviving person %s: %w", survivingID, err)
			}
		}
	}

	// Delete the retiring person; cascade rules handle related records.
	if err := s.persons.Delete(ctx, tx, retiringID); err != nil {
		return fmt.Errorf("person_service: deleting retiring person %s: %w", retiringID, err)
	}

	slog.DebugContext(ctx, "merged persons", "surviving_id", survivingID, "retiring_id", retiringID)
	return nil
}

// PersistDNsForPerson saves any distinguished names that the person does not
// already have.
func (s *PersonService) PersistDNsForPerson(ctx context.Context, tx *sql.Tx, personID string, dnList []string) error {
	for _, dn := range dnList {
		exists, err := s.dns.ExistsByPersonAndDN(ctx, personID, dn)
		if err != nil {
			return fmt.Errorf("person_service: checking DN existence for person %s: %w", personID, err)
		}
		if exists {
			continue
		}
		d := &model.PersonDN{PersonID: personID, DN: dn}
		if err := s.dns.Save(ctx, tx, d); err != nil {
			return fmt.Errorf("person_service: saving DN for person %s: %w", personID, err)
		}
	}
	return nil
}

func optionalString(m map[string]any, key string) *string {
	v, ok := m[key]
	if !ok {
		return nil
	}
	s, _ := v.(string)
	if s == "" {
		return nil
	}
	return &s
}
